File: Simpson_BTK.py
import math
import time
import numpy as np
from scipy.integrate import quad
#Calculate parameters
def calculate_parameters(E,Delta,Gama,Z,T):
    AN = []
    BN = []
    BP = []
    F_E = []
    for number in range(len(E)):
        egama2 = pow(complex(E[number],-Gama),2)
        Delta2 = pow(Delta,2)
        energy_complex = pow((egama2 - Delta2)/egama2,0.5) 
        energy = energy_complex.real
        u02 = (0.5 * (1 + energy))            # coefficient of BCS u0^2
        v02 = (0.5 * (1 - energy))            # coefficient of BCS v0^2
        gama2 = (u02 + Z**2*(u02 - v02))**2
        if abs(complex(E[number],-Gama)) <= Delta:
            AN.append(Delta2/(egama2 + (Delta2 - egama2)*(1+2*Z**2)**2))
            AN[number] = AN[number].real
            BN.append(1 - AN[number])
            BP.append(1)
        else:
            AN.append(u02 * v02/gama2)
            BN.append(pow(u02 - v02,2) * Z*Z * (1 + Z*Z)/gama2)
            BP.append(BN[number])
        F_E.append(math.exp(11.5942*E[number]/T))
    return AN,BN,BP,F_E

# Cut Energy
def split_E(a,n,h):
    E = []                                           # divide the interval
    for i in range(n+1) :
        E.append(a+h*i)
    return E
# Simpson Integration
def integrate_I(V,E,AN,BN,BP,F_E,P,h,T):
    I = []
    time2_start = time.time()
    for m in range(len(V)):
        sum_f_even = 0
        sum_f_odd = 0
        F_V = math.exp(-11.5942*V[m]/T)
        for numberE in range(len(E)):
            # tunneling conductance
            f_e = 1.0/(1+F_E[numberE]*F_V) - 1.0/(1+F_E[numberE])
            f = ((1-P) * (1 + AN[numberE] - BN[numberE]) + P * (1 - BP[numberE]))*f_e
            if numberE == 0:
                f0 = f
            elif numberE == (len(E)-1):
                fn = f
            elif numberE%2 == 0:
                sum_f_even = sum_f_even + f
            else:
                sum_f_odd =  sum_f_odd + f
        I_complex = (h/3)*(f0+fn+2*sum_f_even+4*sum_f_odd)
        I.append(I_complex.real)
    time2_end = time.time()
    print("Integration time:",(time2_end-time2_start),'s')
    return I

# Main Function
def BTK_Diff(parameters,V,T):
    a= -20
    b= 20
    Delta= parameters[0]
    Gama= parameters[1]
    Z= parameters[2]
    P= parameters[3]
    npanel= 1000
    print('Superconducting Gap:' + str(Delta))
    print('Gama:' + str(Gama))
    print('Barrier Height:' + str(Z))
    print('Spin Polarization:' + str(P))
    time_start=time.time()
    #Temperature 
    n = 2 * npanel + 1                               # total number of nodes
    h = (b-a)/n                                      # stepsize
    E = split_E(a,n,h)
    AN,BN,BP,F_E= calculate_parameters(E,Delta,Gama,Z,T)
    I = integrate_I(V,E,AN,BN,BP,F_E,P,h,T)
    dI = []
    dV = []
    G = []
    #if m+1 < len(V):
    #    dV.append(V[m+1] -  V[m])
    for numberV in range(len(V)-1):
        dI.append(I[numberV+1]-I[numberV])
    for numberdI in range(len(dI)):
    	#G.append(round(dI[numberdI]/dV[numberdI],3))
        G.append(dI[numberdI]/dI[numberV]) 
    G.insert(0,1)
    #print(G)
    time_end=time.time()
    print('totally cost:',time_end-time_start,'s')
    return G

File: Simpson_BTK.c
/*
 * ============================================================================
 *
 *       Filename:  Simpson_BTK.c
 *
 *    Description:  Simpson_BTK function C wrapper for python.
 *
 *        Version:  1.0
 *        Created:  2019/01/19 17:27
 *
 *         Author:  Xia Chen
 *          Email:  xiachen@itp.ac.cn
 *
 * ============================================================================
 */

#include <stdlib.h>
#include <math.h>
#include <complex.h>

void c_BTK_Diff(double E[], int NE, double V[], int NV,
                double T, double Delta, double Gama,
                double Z, double P, double h, double G[])
{
    /* Calculate parameters */
    double *AN = (double *)malloc(sizeof(double)*NE);
    double *BN = (double *)malloc(sizeof(double)*NE);
    double *BP = (double *)malloc(sizeof(double)*NE);
    double *F_E = (double *)malloc(sizeof(double)*NE);
    complex double e_gama, egama2;
    double energy, Delta2, u02, v02, gama2;
    for (int i = 0; i < NE; ++i) {
        e_gama = E[i] - Gama * I;
        egama2 = e_gama * e_gama;
        Delta2 = Delta * Delta;
        energy = creal(csqrt((egama2 - Delta2) / egama2));
        u02 = 0.5 * (1 + energy);
        v02 = 0.5 * (1 - energy);
        gama2 = pow(u02 + Z*Z*energy, 2);
        if (cabs(e_gama) <= Delta) {
            AN[i] = creal(Delta2 / (egama2 + (Delta2 - egama2)*pow(1+2*Z*Z, 2)));
            BN[i] = 1 - AN[i];
            BP[i] = 1;
        } else {
            AN[i] = u02 * v02 / gama2;
            BN[i] = energy*energy * Z*Z * (1 + Z*Z)/gama2;
            BP[i] = BN[i];
        }
        F_E[i] = exp(11.5942*E[i]/T);
    }

    /* Integrate */
    double *II = (double *)malloc(sizeof(double)*NV);
    for (int i = 0; i < NV; ++i) {
        double sum_f_even = 0;
        double sum_f_odd = 0;
        double f0 = 0;
        double fn = 0;
        double F_V = exp(-11.5942*V[i]/T);
        for (int j = 0; j < NE; ++j) {
            double f_e = 1.0/(1+F_E[j]*F_V) - 1.0/(1+F_E[j]);
            double f = ((1-P) * (1 + AN[j] - BN[j]) + P*(1-BP[j]))*f_e;
            if (j == 0) f0 = f;
            else if (j == NE - 1) fn = f;
            else if (j%2 == 0) sum_f_even += f;
            else sum_f_odd += f;
        }
        II[i] = (h/3)*(f0+fn + 2*sum_f_even + 4*sum_f_odd);
    }

    /* Differential */
    G[0] = 1;
    double dInormal = II[NV-1] - II[NV-2];
    for (int i = 0; i < NV-1; ++i)
        G[i+1] = (II[i+1]-II[i]) / dInormal;

    free(AN);
    free(BN);
    free(BP);
    free(F_E);
    free(II);
}

// Linux: gcc -Wall -shared -O3 -lm -o libSimpson_BTK.so Simpson_BTK.c
